src/util/expr.h

```diff
@@ -318,6 +318,40 @@ class exprt:public irept
   const_unique_depth_iteratort unique_depth_cend() const;
 };
 
+/// Protected class for all expressions.
+/// This API will eventually replace exprt.
+class protected_exprt : public exprt
+{
+protected:
```

Comment:
This would circumvent the need for this class entirely and just update the classes that inherit from `exprt` to protected extend it and in fact allow for explicitly identifying the classes that use the wrong interface.

Reply:
There are a lot of classes inheriting from `exprt`; furthermore, the interface of `exprt` should change as well, once we are there.